### ISSUE TITLE
Some questions about language model

### ISSUE THREAD

**Issue by WanderingSven** [espnet/espnet#3521]
Hello everyone:
          I am using the espnet1 to train RNNLM with 460h text, and I use both 960h text and librispeech-lm-norm to train transformer LM. The configures as follow:
# Transformer configure:
model-module: transformer
att-unit: 512
embed-unit: 128
head: 8 
layer: 12
pos-enc: none
unit: 2048

minibatch related
batchsize: 64
maxlen: 40

optimization related
opt: adam
schedulers: lr=cosine
dropout-rate: 0.0
epoch: 50
gradclip: 1.0
lr: 1e-4
lr-cosine-total: 100000
lr-cosine-warmup: 1000
patience: 0
sortagrad: 0

# RNN configure:
layer: 2
unit: 1024
opt: adam      # or adam
sortagrad: 0   # Feed samples from shortest to longest ; -1: enabled for all epochs, 0: disabled, other: enabled for 'other' epochs
batchsize: 128  # batch size in LM training
epoch: 60      # if the data size is large, we can reduce this
patience: 0
maxlen: 80     # if sentence length > lm_maxlen, lm_batchsize is automatically reduced
dropout-rate: 0.3
lr: 0.0005
#

However, I find the val perplexity of RNNLM reaches 26.x after 3 epoch while the val perplexity of transformer LM is 33.x. The logs are as follow:
# RNNLM log:
"main/loss": 180.13677032470704,
 "main/nll": 23057.5066015625,
 "main/count": 5656.36,
  "validation/main/loss": 1872.0366592407227,
  "validation/main/nll": 227094.93408203125,
  "validation/main/count": 69469.0,
  "epoch": 3,
  "iteration": 1600,
  "elapsed_time": 407.69571410119534,
"perplexity": 58.932102724877495
  "val_perplexity": 26.28533366850296
# Transformer log:
"main/loss": 3.527453274726868,
"main/nll": 4767.065887451172,
"main/count": 1356.51,
"validation/main/loss": 179.45932841300964,
"validation/main/nll": 244071.1156616211,
"validation/main/count": 69469.0,
"epoch": 7,
"iteration": 1346100,
"elapsed_time": 199941.5584934177,
"perplexity": 33.58950267355802,
"val_perplexity": 33.56157239489766
#  
I used more training data and epoch numbers on transformer LM while the result is worse. Is this result reasonable?

**Comment by sw005320** [comment]
Some overtraining seems to happen in your case since transformer training perplexity is much better than RNNLM.
Can you tune the learning rate or other regularization-related hyper-parameters (like dropout)?

**Comment by WanderingSven** [comment]
@sw005320 Thanks for your reply.  In [espnet/nets/pytorch_backbend/lm/default](https://github.com/espnet/espnet/blob/51e3299fb706bd2860afae6ef2f5dcb8fbb0f67b/espnet/nets/pytorch_backend/lm/default.py#L135):
`loss += loss_batch.mean() * non_zeros
`
if loss batch=[2,4,0], non zeros=3 then loss batch.mean()=2,loss=6, however, the true loss=6. Why do we need to calculate the mean of loss_batch? I think it will cause the wrong ppl.

**Comment by WanderingSven** [comment]
@sw005320 Thanks for your reply.  In [espnet/nets/pytorch_backbend/lm/default](https://github.com/espnet/espnet/blob/51e3299fb706bd2860afae6ef2f5dcb8fbb0f67b/espnet/nets/pytorch_backend/lm/default.py#L135):
`loss += loss_batch.mean() * non_zeros
`
if loss batch=[2,4,0], non zeros=2 then loss batch.mean()=2,loss=4, however, the true loss=6. Why do we need to calculate the mean of loss_batch? I think it will cause the wrong ppl.
